crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -574,13 +577,21 @@ where
             }
             ast::Stmt::AnnAssign(node) => {
                 debug_assert!(self.current_assignment.is_none());
-                self.visit_expr(&node.annotation);
-                if let Some(value) = &node.value {
-                    self.visit_expr(value);
+                let valid_target = matches!(
+                    *node.target,
+                    Expr::Attribute(_) | Expr::Subscript(_) | Expr::Name(_)
+                );
+                if valid_target {
```

Comment:
if we visit with an invalid target we get downstream errors on re-defining things. There's likely a way to visit here with an invalid target that doesn't generate issues